pyccel/ast/builtins.py

```diff
@@ -1045,7 +1062,7 @@ def __init__(self, *x):
 
         if not x.is_homogeneous:
             types = ', '.join(str(xi.dtype) for xi in x)
-            raise PyccelError("Cannot determine final dtype of 'max' call with arguments of different "
+            raise TypeError("Cannot determine final dtype of 'max' call with arguments of different "
```

Comment:
Why have you decided not to use `PyccelError` any more?

Reply:
In Pyccel we have 3 types of error:
- `PyccelSyntaxError`
- `PyccelSemanticError`
- `PyccelCodegenError`

each of these inherits from `PyccelError`. They are raised by `errors.report` and caught in `codegen/pipeline.py`. Ideally we want to avoid doing `raise PyccelError` directly because it means that the error is caught in pipeline.py without any context (no symbol, no line number, etc).

Before this PR we were using `PyccelError` in builtin functions because there was no try/except call around these calls. After this PR these functions are handled in the same way as imported functions (e.g. `MathCos`). for these functions we catch `TypeErrors` here:
https://github.com/pyccel/pyccel/blob/046e93f1ca42839698de08de6ab08beeb2f0943d/pyccel/parser/semantic.py#L1058-L1065

this allows the correct subclass of error to be raised with the necessary context.

I've just checked and there is now only 1 `PyccelError` which is raised directly. I have removed it too: 65e9698c1

Now that that is done I think it may be worth changing our approach in the future. Instead of raising `TypeError`s we should probably be raising `PyccelError`s

Reply:
Thanks for the detailed explanation!

I agree that raising `TypeError`s all the time is probably not ideal. This calls for a separate discussion.